alacritty_terminal/src/tty/unix.rs

```diff
@@ -121,17 +118,33 @@ impl Pty {
     }
 }
 
-#[cfg(target_os = "macos")]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    let shell_name = pw.shell.rsplit('/').next().unwrap();
-    let argv = vec![String::from("-c"), format!("exec -a -{} {}", shell_name, pw.shell)];
-
-    Program::WithArgs { program: "/bin/bash".to_owned(), args: argv }
+/// Look for a shell in the `$SHELL` environment variable, then in `passwd`.
+fn default_shell(pw: &Passwd<'_>) -> String {
+    env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned())
 }
 
 #[cfg(not(target_os = "macos"))]
-fn default_shell(pw: &Passwd<'_>) -> Program {
-    Program::Just(env::var("SHELL").unwrap_or_else(|_| pw.shell.to_owned()))
+fn default_shell_command(pw: &Passwd<'_>) -> Command {
+    Command::new(default_shell(pw))
+}
+
+#[cfg(target_os = "macos")]
+fn default_shell_command(pw: &Passwd<'_>) -> Command {
+    let shell = default_shell(pw);
+    let shell_name = shell.rsplit('/').next().unwrap();
+
+    // On macOS, use the `login` command so the shell will appear as a tty session.
+    let mut login_command = Command::new("/usr/bin/login");
+
+    // Exec the shell with argv[0] prepended by '-' so it becomes a login shell.
+    // `login` normally does this itself, but `-l` disables this.
+    let exec = format!("exec -a -{} {}", shell_name, shell);
+
+    // -f: Bypasses authentication for the already-logged-in user.
+    // -l: Skips changing directory to $HOME and prepending '-' to argv[0].
+    // -p: Preserves the environment.
+    login_command.args(["-flp", pw.name, "/bin/sh", "-c", &exec]);
+    login_command
```

Comment:
Reads like a novel by Charles Dickens, much nicer now. :+1:

Reply:
More like a greek tragedy in 3 acts...